### ISSUE TITLE
An io_uring based Transport Layer (Part III) - Preview Release

### ISSUE THREAD

**Issue by tkp1n** [tkp1n/ndportmann.com#2]
Issue for feedback and discussion around the post [An io_uring based Transport Layer (Part III) - Preview Release](https://ndportmann.com/io_uring-preview-release)

**Comment by onanur993** [comment]
@tkp1n 
How can io_uring submission queue size be set when the number of sockets is dynamic? I want to have one receive and one send per socket.

**Comment by tkp1n** [comment]
Good question, @ona993!
I've asked myself the same question when implementing IoUring.Transport and came up with two potential solutions. For both solutions you would want to set the ring size large enough to be able to handle the expected number of concurrent connections/sockets. Even with the best guess, you need an overflow strategy: 

1. Switch to a bigger ring, once you've reached the limits of the current instance. This is similar to how C++ vectors or C# Lists grow when needed.
2. Maintain a queue-like data structure to store all the submissions you'd like to submit but no longer fit the ring. Make sure to process the items in this queue directly after submitting to the kernel to avoid operations getting stalled for too long in this queue.

The second approach is also how the kernel handles completion queue overflows on kernels that report `IORING_FEAT_NODROP`. 
In IoUring.Transport I've also chosen this second approach as I liked the symmetry to what the kernel is doing.

**Comment by jsoneaday** [comment]
This is very cool. Do you know if the windows kernel will get this io capability eventually?

**Comment by tkp1n** [comment]
> This is very cool. Do you know if the windows kernel will get this io capability eventually?

Glad you like it. 😃
On the Windows side of things, we have IOCP allowing us to execute multiple asynchronous IO operations "at once". IOCP is rather old but still surprisingly competitive compared to the Linux alternatives (at least prior to io_uring). I'd be surprised if Windows gets another set of IO APIs any time soon.

**Comment by jsoneaday** [comment]
Yes I did hear that IOCP was originally faster than Linux, at least until now.

Great work! Looking forward
________________________________
